File: src/Solutions/Euler260.hs
module Solutions.Euler260 where
import Utils.Operators((<&&>))
import Control.Monad
import Data.Array.ST
import Data.Array.Unboxed

type I = (Int, Int, Int)

isSorted :: I -> Bool
isSorted (a, b, c) = a >= b && b >= c
sumI :: I -> Int
sumI (a, b, c) = a + b + c

sort :: I -> I
sort (a, b, c)
    | a < b =
        if b < c
            then (c, b, a)
            else if a < c
                then (b, c, a)
                else (b, a, c)
    | b < c =
            if a < c
                then (c, a, b)
                else (a, c, b)
    | otherwise = (a, b, c)

run :: Int -> UArray I Bool
run n = runSTUArray $ do
    arr <- newArray ((0, 0, 0), (n, n, n)) False -- True is winning position, False is losing
    forM_ [0..n] $ \x -> forM_ [0..x] $ \y -> forM_ [0..y] $ \z -> -- x >= y >= z
        readArray arr (x, y, z) >>= \p -> unless p $ do
            forM_ [1..n - x] $ \d -> writeArray arr (x + d, y, z) True
            forM_ [1..n - y] $ \d -> writeArray arr (sort (x, y + d, z)) True
            forM_ [1..n - z] $ \d -> writeArray arr (sort (x, y, z + d)) True
            forM_ [1..n - x] $ \d -> writeArray arr (x + d, y + d, z) True
            forM_ [1..n - x] $ \d -> writeArray arr (sort (x + d, y, z + d)) True
            forM_ [1..n - y] $ \d -> writeArray arr (sort (x, y + d, z + d)) True
            forM_ [1..n - x] $ \d -> writeArray arr (x + d, y + d, z + d) True
    return arr

f = sum . map (sumI . fst) . filter (not . snd <&&> isSorted . fst) . assocs . run

euler260 :: IO String
euler260 = return $ show $ f 1000

File: rust/src/solutions/euler260.rs
fn sort(arr: [usize; 3]) -> [usize; 3] {
    let mut x = arr.clone();
    x.sort();
    x.reverse();
    x
}

pub fn euler260() -> String {
    let n = 1000;
    let mut arr = vec![vec![vec![false; n + 1]; n + 1]; n + 1];
    for x in 0..=n {
        for y in 0..=x {
            for z in 0..=y {
                if !arr[x][y][z] {
                    for d in 1..=n - x {
                        arr[x + d][y][z] = true;
                    }
                    for d in 1..=n - y {
                        let [xx, yy, zz] = sort([x, y + d, z]);
                        arr[xx][yy][zz] = true;
                    }
                    for d in 1..=n - z {
                        let [xx, yy, zz] = sort([x, y, z + d]);
                        arr[xx][yy][zz] = true;
                    }
                    for d in 1..=n - x {
                        arr[x + d][y + d][z] = true;
                    }
                    for d in 1..=n - x {
                        let [xx, yy, zz] = sort([x + d, y, z + d]);
                        arr[xx][yy][zz] = true;
                    }
                    for d in 1..=n - y {
                        let [xx, yy, zz] = sort([x, y + d, z + d]);
                        arr[xx][yy][zz] = true;
                    }
                    for d in 1..=n - x {
                        arr[x + d][y + d][z + d] = true;
                    }
                }
            }
        }
    }
    let mut ans = 0;
    for x in 0..=n {
        for y in 0..=x {
            for z in 0..=y {
                if !arr[x][y][z] {
                    ans += x + y + z;
                }
            }
        }
    }
    ans.to_string()
}
